### PR TITLE
[aes] Use one-hot encoding for OPERATION field in main control register

### PR DESCRIPTION
**Pull request by vogelpi**
Sorry, it got a bit more involved than anticipated as invalid encodings need to be catched inside the cipher core.

@msfschaffner, @tjaychen, @cdgori can you please take a look? We should get this in as it's not transparent DV and therefore blocks other DV tasks.

This is related to lowRISC/OpenTitan#10422.

### CODE REVIEW

File: sw/device/lib/dif/dif_aes.c [hunk 1]
```diff
@@ -111,6 +123,38 @@ static void aes_clear_internal_state(const dif_aes_t *aes) {
...
+    case kDifAesOperationEncrypt:
+      return kAesOperationFieldValEnc;
+    case kDifAesOperationDecrypt:
+      return kAesOperationFieldValDec;
+    default:
+      return kAesOperationFieldValInvalid;
+  }
+}
+
+// TODO: We should extract the MODE from the transcation using this function and
```
Comment (msfschaffner) [inline sw/device/lib/dif/dif_aes.c:138]: nittynit: transcation -> transaction

**Review by msfschaffner — APPROVED** [review]
LGTM, but would be good to get some SW feedback on the DIF changes.

File: sw/device/lib/dif/dif_aes_unittest.cc [hunk 2]
```diff
@@ -60,9 +60,10 @@ class AesTestInitialized : public AesTest {
   dif_aes_t aes_;
 
   const dif_aes_transaction_t kTransaction = {
+      .operation = kDifAesOperationEncrypt,
+      .mode = kDifAesModeEcb,
       .key_len = kDifAesKey128,
-      .mode = kDifAesModeEncrypt,
-      .operation = kDifAesOperationAuto,
+      .manual_operation = kDifAesManualOperationAuto,
```
Comment (timothytrippel) [inline sw/device/lib/dif/dif_aes_unittest.cc:66]: Maybe it's better to have this be a `dif_toggle_t` type, rather than a separate enum? Then this field could just be whether manual operation is "enabled" or "disabled". As is, it a tad confusing to have two enum types `*_operation_t` and `*_manual_operation_`.

**Review by timothytrippel — APPROVED** [review]
Minor comment on DIF, otherwise LGTM.

File: hw/ip/aes/rtl/aes_cipher_control_fsm.sv [hunk 3]
```diff
@@ -226,7 +227,8 @@ module aes_cipher_control_fsm import aes_pkg::*;
             // Load full key
             key_full_sel_o = dec_key_gen_i ? KEY_FULL_ENC_INIT :
                         (op_i == CIPH_FWD) ? KEY_FULL_ENC_INIT :
-                                             KEY_FULL_DEC_INIT;
+                        (op_i == CIPH_INV) ? KEY_FULL_DEC_INIT :
+                                             KEY_FULL_ENC_INIT;
```
Comment (tjaychen) [inline hw/ip/aes/rtl/aes_cipher_control_fsm.sv:231]: just curious honestly, why `ENC` when no match? is this better than using `DEC`?

File: hw/ip/aes/rtl/aes_cipher_control_fsm.sv [hunk 4]
```diff
@@ -304,7 +306,8 @@ module aes_cipher_control_fsm import aes_pkg::*;
         prng_reseed_req_o = Masking & prng_reseed_q_i & ~prng_reseed_done_q;
 
         // Select round key: direct or mixed (equivalent inverse cipher)
-        round_key_sel_o = (op_i == CIPH_FWD) ? ROUND_KEY_DIRECT : ROUND_KEY_MIXED;
+        round_key_sel_o = (op_i == CIPH_FWD) ? ROUND_KEY_DIRECT :
+                          (op_i == CIPH_INV) ? ROUND_KEY_MIXED  : ROUND_KEY_DIRECT;
```
Comment (tjaychen) [inline hw/ip/aes/rtl/aes_cipher_control_fsm.sv:310]: since the above picked `ENC`, does it make sense for this to pick something that is not related to `ENC` when the incoming operation is invalid?

File: hw/ip/aes/rtl/aes_cipher_control_fsm_p.sv [hunk 5]
```diff
@@ -150,31 +153,32 @@ module aes_cipher_control_fsm_p import aes_pkg::*;
...
-  logic             key_expand_out_req;
-  logic [3:0]       rnd_ctr_q;
-  logic [3:0]       rnd_ctr_rem_q;
-  logic [3:0]       num_rounds_q;
-  logic             crypt_q;
-  logic             dec_key_gen_q;
-  logic             prng_reseed_q;
-  logic             key_clear_q;
-  logic             data_out_clear_q;
+  logic                 in_valid;
```
Comment (tjaychen) [inline hw/ip/aes/rtl/aes_cipher_control_fsm_p.sv:156]: is there a part of you that's thinking "man i wish i put this in a struct"... i find myself thinking that a lot when I'm plumbing a signal through a hierarchy.. hehe

**Review by tjaychen — APPROVED** [review]
lgtm, just a couple of minor questions.

File: hw/ip/aes/rtl/aes_cipher_control_fsm.sv (hunk 3 above)
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_cipher_control_fsm.sv:231]: The rationale here is that in most application scenarios the cipher will be operated in forward direction anyways. Inverse operation is needed only for ECB and CBC decryption. By using `ENC` as default here, we can thus make it harder for an adversary to manipulate the behavior in most cases. But it's a bit hypothetical, as a manipulated `op_i` should already be caught by the detector.

File: hw/ip/aes/rtl/aes_cipher_control_fsm.sv (hunk 4 above)
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_cipher_control_fsm.sv:310]: This is valid point. However, in most application scenarios `op_i` will be `CIPH_FWD` (see previous point). Adding a third MUX input here would also be pretty costly: it's 2x 128 bit wide.

File: hw/ip/aes/rtl/aes_cipher_control_fsm_p.sv (hunk 5 above)
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_cipher_control_fsm_p.sv:156]: ;-) I fully agree with you Tim. I hope to have some time after ES to automate part of this. It would make the adoption of the multirail countermeasure in other IP more feasible.

File: sw/device/lib/dif/dif_aes_unittest.cc (hunk 2 above)
Comment (vogelpi) [reply sw/device/lib/dif/dif_aes_unittest.cc:66]: This sounds indeed like an option. But I am wondering if it isn't simpler to keep it following the hardware. Anyway, I've added your proposal to corresponding issue #10487. I will leave it as is for now to move forward and unblock DV folks waiting on this PR.

**Review by vogelpi — COMMENTED** [review]
Thanks everybody for your fast reviews!

File: sw/device/lib/dif/dif_aes.c (hunk 1 above)
Comment (vogelpi) [reply sw/device/lib/dif/dif_aes.c]: This is fixed now, thanks.

File: hw/ip/aes/rtl/aes_cipher_control_fsm.sv (hunk 4 above)
Comment (tjaychen) [reply hw/ip/aes/rtl/aes_cipher_control_fsm.sv:310]: o i just meant changing the last item from `ROUND_KEY_DIRECT` to `ROUND_KEY_MIXED`.  That should not add more logic than what is already here right..? It's a very minor point, I was just thinking if invalid, we might as well do something that does not really make sense.

File: hw/ip/aes/rtl/aes_cipher_control_fsm.sv (hunk 3 above)
Comment (cdgori) [reply hw/ip/aes/rtl/aes_cipher_control_fsm.sv:231]: Good explanation of your thinking, even if it is all a bit theoretical as you say.

File: hw/ip/aes/rtl/aes_cipher_control_fsm.sv (hunk 4 above)
Comment (cdgori) [reply hw/ip/aes/rtl/aes_cipher_control_fsm.sv:310]: I think I agree with @tjaychen here.  The MIXED key should be OK to use here and makes it even harder to manipulate the block successfully/usefully.

**Review by cdgori — COMMENTED** [review]
post-merge but LGTM.  I am sort of amazed at the number of files this touched, but it's a good change to make.

File: hw/ip/aes/rtl/aes_cipher_control_fsm.sv (hunk 4 above)
Comment (vogelpi) [reply hw/ip/aes/rtl/aes_cipher_control_fsm.sv:310]: Ah, I forgot to reply here, sorry Tim. I think I will do a follow-up PR to switch this to MIXED then.

Comment (tjaychen) [reply hw/ip/aes/rtl/aes_cipher_control_fsm.sv:310]: honestly I don't think it's a huge deal, if you're tied up with other things this is not that critical.
@cdgori do you agree?

Comment (vogelpi) [reply hw/ip/aes/rtl/aes_cipher_control_fsm.sv:310]: No worries, this would be a simple change and I planned to do this early next week together with some other simple, FI-related changes.
 
On the other hand, I just realized that this actually increases area whenever the cipher is instantiated in a module where only FWD operation is needed (e.g. CSRNG). So, I am again inclined not to do this change. Actually, I think I need to rework this again slightly. I just realized that since I made merge this PR, area of CSRNG increased by 5 kGE. The reason is simply that I've put a `prim_buf` on the `op_i` input which completely defeats constant propagation (in CSRNG we constantly set `op_i ( aes_pkg::AES_ENC ),`. Thanks for starting this discussion @tjaychen and @cdgori !

Comment (vogelpi) [reply hw/ip/aes/rtl/aes_cipher_control_fsm.sv:310]: FYI @mwbranstad if you wonder why CSRNG suddenly increased in size this week.

Comment (cdgori) [reply hw/ip/aes/rtl/aes_cipher_control_fsm.sv:310]: Agree with @tjaychen it's not critical.  And in fact, your point about FWD-only operation is a good one so let's not worry about this.  (You'll have to figure out the constant propagation issue though.)

Comment (vogelpi) [reply hw/ip/aes/rtl/aes_cipher_control_fsm.sv:310]: Alright, I've filed a PR here to re-enable constant propagration inside CSRNG: #10717
